codechecker_common/checker_labels.py

```diff
@@ -0,0 +1,219 @@
+        labels = data['labels']
+
+        if not isinstance(constraints, dict) or not isinstance(labels, dict):
+            raise ValueError(
+                f'{header} "constraints" and "labels" should be assigned JSON '
+                'objects.')
+
+        for label, properties in constraints.items():
+            if not isinstance(properties, dict):
+                raise ValueError(
+                    f'{header} Constraints of "{label}" should be described '
+                    'by a JSON object.')
+
+            for property, value in properties.items():
+                if property == 'choice' and not isinstance(value, dict):
+                    raise ValueError(
+                        f'{header} label "choice" should be assigned a JSON '
+                        'object.')
+                elif property == 'required' and not isinstance(value, bool):
+                    raise ValueError(
+                        f'{header} label "required" should be assigned a '
+                        'boolean value.')
+                elif property == 'unique' and not isinstance(value, bool):
+                    raise ValueError(
+                        f'{header} label "unique" should be assigned a '
+                        'boolean value.')
+                elif property == 'default' and not isinstance(value, str):
+                    raise ValueError(
+                        f'{header} label "default" should be assigned a '
+                        'string value.')
+
+            if properties.get('default') and not properties.get('unique'):
+                raise ValueError(
+                    f'{header} "default" value can belong to labels only '
+                    'with "unique" constraint.')
+
+        for checker, labels in labels.items():
+            if not is_string_list(labels):
+                raise ValueError(
+                    f'{header} "{checker}" should be assigned a string list.')
+
+            if any(map(lambda s: ':' not in s, labels)):
+                raise ValueError(
+                    f'{header} all labels at "{checker}" should be in the '
+                    'following format: <label>:<property>.')
+
+            check_constraints(checker, labels)
+
+    def checkers_by_labels(self, filter_labels):
+        """
+        Returns a list of checkers that have at least one of the specified
+        labels.
+
+        filter_labels -- A string list which contains labels with specified
+                         values. E.g. ['profile:default', 'severity:high'].
+        """
+        result = []
+
+        filter_labels = set(map(self.__get_label_key_value, filter_labels))
+
+        for checker, labels in self.__data['labels'].items():
+            labels = set(map(self.__get_label_key_value, labels))
+
+            if labels.intersection(filter_labels):
+                result.append(checker)
+
+        return result
+
+    def label_of_checker(self, checker, label):
```

Comment:
I'm not so sure about this. Labels are rather untyped things. It is just a historical thing that we use severity and profile for a long type. As for avoiding typos, listing labels in the "constraints" section is mandatory, so it could be checked if a valid label name was given here.

Reply:
Done